controllers/etcd_controller_test.go

```diff
@@ -54,85 +90,229 @@ var _ = Describe("Druid", func() {
+				p.Status.ContainerStatuses = []corev1.ContainerStatus{
+					{
+						Name: "Container-0",
+						State: corev1.ContainerState{
+							Waiting: &corev1.ContainerStateWaiting{
+								Reason:  "CrashLoopBackOff",
+								Message: "Container is in CrashLoopBackOff.",
+							},
+						},
+					},
+				}
+				err = c.Status().Update(context.TODO(), p)
+				Expect(err).NotTo(HaveOccurred())
+				storeSecret := instance.Spec.Backup.Store.SecretRef.Name
+				errors := createSecrets(c, instance.Namespace, storeSecret)
+				Expect(len(errors)).Should(BeZero())
+
+			})
+			It("should restart pod", func() {
+				defer WithWd("..")()
+				err = c.Create(context.TODO(), instance)
+				Expect(err).NotTo(HaveOccurred())
+				Eventually(func() error { return podDeleted(c, instance) }, timeout, pollingInterval).Should(BeNil())
+			})
+			AfterEach(func() {
+				s := &appsv1.StatefulSet{}
+				Eventually(func() error { return statefulsetIsCorrectlyReconcilced(c, instance, s) }, timeout, pollingInterval).Should(BeNil())
+				setStatefulSetReady(s)
+				err = c.Status().Update(context.TODO(), s)
+				Expect(err).NotTo(HaveOccurred())
+				c.Delete(context.TODO(), instance)
+			})
 		})
 	})
 
+	DescribeTable("when deleting etcd resources",
+		func(name string, setupStatefulSet StatefulSetInitializer, setupPod bool) {
+			var ss *appsv1.StatefulSet
+			var sts appsv1.StatefulSet
+			instance := getEtcd(name, "default", false)
+			c := mgr.GetClient()
+			switch setupStatefulSet {
+			case WithoutOwnerReference:
+				ss = createStatefulset(name, "default", instance.Spec.Labels)
+				Expect(ss.OwnerReferences).Should(BeNil())
+			case WithOwnerReference:
+				ss = createStatefulsetWithOwnerReference(instance)
+				Expect(len(ss.OwnerReferences)).ShouldNot(BeZero())
+			case WithOwnerAnnotation:
+				ss = createStatefulsetWithOwnerAnnotations(instance)
+			default:
+				Fail("StatefulSetInitializer invalid")
+			}
+
+			storeSecret := instance.Spec.Backup.Store.SecretRef.Name
+			errors := createSecrets(c, instance.Namespace, storeSecret)
+			Expect(len(errors)).Should(BeZero())
+			c.Create(context.TODO(), ss)
+			defer WithWd("..")()
+			err := c.Create(context.TODO(), instance)
+			Expect(err).NotTo(HaveOccurred())
+			Eventually(func() error { return statefulsetIsCorrectlyReconcilced(c, instance, &sts) }, timeout, pollingInterval).Should(BeNil())
+			setStatefulSetReady(&sts)
+			err = c.Status().Update(context.TODO(), &sts)
+			Expect(err).NotTo(HaveOccurred())
+			err = c.Delete(context.TODO(), instance)
+			Expect(err).NotTo(HaveOccurred())
+			Eventually(func() error { return statefulSetRemoved(c, &sts) }, timeout, pollingInterval).Should(BeNil())
+			Eventually(func() error { return etcdRemoved(c, instance) }, timeout, pollingInterval).Should(BeNil())
+		},
+		Entry("when  statefulset with ownerReference and without owner annotations, druid should adopt and delete statefulset", "foo61", WithOwnerReference, false),
```

Comment:
Should we have a case of pre-existing `StatefulSet` with no ownerReference and no annotations in which case the it should not be removed?

Reply:
Actually, if there is no owner reference/ owner annotations, the druid will adopt it if the selector match. This is because, in the pre druid scenario, we didnt have both and druid went ahead adopting it. Should we not continue the behavior?

Reply:
But it won't adopt if the etcd being deletes, no? Then the `StatefulSet` would be left untouched. A negative case.

Reply:
Can you please make the change for the negative case, I mentioned above? I did not mean the adoption+deletion case.

Reply:
Based on offline conversation. We can take it forward to another PR since this requires reconcile call to be brought in into the test func.